File: Data Structures & Algorithms/house-robber-ii/submission-1.kt
class Solution {
    fun rob(nums: IntArray): Int {
        if (nums.size == 1) return nums[0]
        val memo = Array(nums.size) { IntArray(2) { -1 } }

        fun dfs(i: Int, flag: Int): Int {
            if (i >= nums.size || (flag == 1 && i == nums.size - 1)) {
                return 0
            }
            if (memo[i][flag] != -1) {
                return memo[i][flag]
            }
            memo[i][flag] = maxOf(dfs(i + 1, flag),
                                  nums[i] + dfs(i + 2, flag or if (i == 0) 1 else 0))
            return memo[i][flag]
        }

        return maxOf(dfs(0, 1), dfs(1, 0))
    }
}

File: Data Structures & Algorithms/jump-game-ii/submission-1.cpp
class Solution {
   public:
    int jump(vector<int>& nums) {
        unordered_map<int, int> memo;
        return dfs(nums, 0, memo);
    }

   private:
    int dfs(vector<int>& nums, int index, unordered_map<int, int>& memo) {
        if (memo.count(index)) {
            return memo[index];
        }
        if (index == nums.size() - 1) {
            return 0;
        }
        if (nums[index] == 0) {
            return 1000000;
        }

        int result = 1000000;
        int end = min((int)nums.size(), index + nums[index] + 1);
        for (int j = index + 1; j < end; j++) {
            result = min(result, 1 + dfs(nums, j, memo));
        }
        memo[index] = result;
        return result;
    }
};
